File: internal/scm/bitbucket/bitbucket.go
package bitbucket

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/CrowdStrike/codestrike/internal/scm"
)

var _ scm.Client = (*Client)(nil)

// Config holds the Bitbucket API connection settings.
type Config struct {
	Workspace string
	RepoSlug  string
	Token     string
	BaseURL   string
}

// Client implements scm.Client for Bitbucket Cloud.
type Client struct {
	config     Config
	httpClient *http.Client
}

// New creates a Bitbucket client with default HTTP client.
func New(cfg Config) *Client {
	if cfg.BaseURL == "" {
		cfg.BaseURL = "https://api.bitbucket.org/2.0"
	}
	return &Client{
		config:     cfg,
		httpClient: http.DefaultClient,
	}
}

// NewWithHTTPClient creates a Bitbucket client with a custom HTTP client (for testing).
func NewWithHTTPClient(cfg Config, httpClient *http.Client) *Client {
	c := New(cfg)
	c.httpClient = httpClient
	return c
}

func (c *Client) PullRequestExists(ctx context.Context, number int) (bool, error) {
	url := fmt.Sprintf("%s/repositories/%s/%s/pullrequests/%d",
		c.config.BaseURL, c.config.Workspace, c.config.RepoSlug, number)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return false, fmt.Errorf("creating request: %w", err)
	}
	c.setAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return false, fmt.Errorf("executing request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return false, nil
	}
	if resp.StatusCode != http.StatusOK {
		return false, fmt.Errorf("unexpected status %d", resp.StatusCode)
	}

	return true, nil
}

func (c *Client) GetPullRequestDescription(ctx context.Context, number int) (string, string, error) {
	url := fmt.Sprintf("%s/repositories/%s/%s/pullrequests/%d",
		c.config.BaseURL, c.config.Workspace, c.config.RepoSlug, number)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return "", "", fmt.Errorf("creating request: %w", err)
	}
	c.setAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", "", fmt.Errorf("executing request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", "", fmt.Errorf("unexpected status %d", resp.StatusCode)
	}

	var pr struct {
		Title   string `json:"title"`
		Summary struct {
			Raw string `json:"raw"`
		} `json:"summary"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&pr); err != nil {
		return "", "", fmt.Errorf("decoding response: %w", err)
	}

	return pr.Title, pr.Summary.Raw, nil
}

func (c *Client) GetPullRequestDiff(ctx context.Context, number int) (string, error) {
	url := fmt.Sprintf("%s/repositories/%s/%s/pullrequests/%d/diff",
		c.config.BaseURL, c.config.Workspace, c.config.RepoSlug, number)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return "", fmt.Errorf("creating request: %w", err)
	}
	c.setAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("executing request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return "", fmt.Errorf("pull request #%d not found", number)
	}
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("unexpected status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("reading response: %w", err)
	}

	return string(body), nil
}

func (c *Client) GetPullRequestFiles(ctx context.Context, number int) ([]scm.PullRequestFile, error) {
	entries, err := c.fetchDiffstat(ctx, number)
	if err != nil {
		return nil, fmt.Errorf("fetching diffstat: %w", err)
	}

	fullDiff, err := c.GetPullRequestDiff(ctx, number)
	if err != nil {
		return nil, fmt.Errorf("fetching diff: %w", err)
	}
	patches := splitDiff(fullDiff)

	files := make([]scm.PullRequestFile, 0, len(entries))
	for _, e := range entries {
		files = append(files, scm.PullRequestFile{
			Filename: e.Path,
			Status:   e.Status,
			Patch:    patches[e.Path],
		})
	}

	return files, nil
}

func (c *Client) GetFileContent(ctx context.Context, path, ref string) (string, error) {
	if ref == "" {
		ref = "HEAD"
	}
	url := fmt.Sprintf("%s/repositories/%s/%s/src/%s/%s",
		c.config.BaseURL, c.config.Workspace, c.config.RepoSlug, ref, path)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return "", fmt.Errorf("creating request: %w", err)
	}
	c.setAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("executing request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return "", fmt.Errorf("file %q not found", path)
	}
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("unexpected status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("reading response: %w", err)
	}

	return string(body), nil
}

func (c *Client) PublishComment(ctx context.Context, number int, body string) error {
	url := fmt.Sprintf("%s/repositories/%s/%s/pullrequests/%d/comments",
		c.config.BaseURL, c.config.Workspace, c.config.RepoSlug, number)

	payload, err := json.Marshal(map[string]any{
		"content": map[string]string{"raw": body},
	})
	if err != nil {
		return fmt.Errorf("marshaling comment: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(payload))
	if err != nil {
		return fmt.Errorf("creating request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	c.setAuth(req)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("executing request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return fmt.Errorf("pull request #%d not found", number)
	}
	if resp.StatusCode != http.StatusCreated {
		return bitbucketAPIError(resp)
	}

	return nil
}

func (c *Client) setAuth(req *http.Request) {
	if c.config.Token != "" {
		req.Header.Set("Authorization", "Bearer "+c.config.Token)
	}
}

type diffstatEntry struct {
	Status string `json:"status"`
	Path   string `json:"-"`
}

type diffstatNewOld struct {
	Path string `json:"path"`
}

type diffstatRaw struct {
	Status string         `json:"status"`
	New    diffstatNewOld `json:"new"`
	Old    diffstatNewOld `json:"old"`
}

func (c *Client) fetchDiffstat(ctx context.Context, number int) ([]diffstatEntry, error) {
	url := fmt.Sprintf("%s/repositories/%s/%s/pullrequests/%d/diffstat",
		c.config.BaseURL, c.config.Workspace, c.config.RepoSlug, number)

	var entries []diffstatEntry
	for url != "" {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		if err != nil {
			return nil, fmt.Errorf("creating request: %w", err)
		}
		c.setAuth(req)

		resp, err := c.httpClient.Do(req)
		if err != nil {
			return nil, fmt.Errorf("executing request: %w", err)
		}

		if resp.StatusCode != http.StatusOK {
			resp.Body.Close()
			return nil, fmt.Errorf("unexpected status %d", resp.StatusCode)
		}

		var page struct {
			Values []diffstatRaw `json:"values"`
			Next   string        `json:"next"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&page); err != nil {
			resp.Body.Close()
			return nil, fmt.Errorf("decoding response: %w", err)
		}
		resp.Body.Close()

		for _, v := range page.Values {
			path := v.New.Path
			if path == "" {
				path = v.Old.Path
			}
			entries = append(entries, diffstatEntry{
				Status: v.Status,
				Path:   path,
			})
		}

		url = page.Next
	}

	return entries, nil
}

func splitDiff(fullDiff string) map[string]string {
	patches := make(map[string]string)
	chunks := strings.Split(fullDiff, "diff --git ")

	for _, chunk := range chunks[1:] {
		newlineIdx := strings.Index(chunk, "\n")
		if newlineIdx == -1 {
			continue
		}
		header := chunk[:newlineIdx]
		parts := strings.Fields(header)
		if len(parts) < 2 {
			continue
		}
		path := strings.TrimPrefix(parts[1], "b/")
		patches[path] = "diff --git " + chunk
	}

	return patches
}

func bitbucketAPIError(resp *http.Response) error {
	var apiError struct {
		Error struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	if err := json.NewDecoder(io.LimitReader(resp.Body, 1<<20)).Decode(&apiError); err == nil && apiError.Error.Message != "" {
		return fmt.Errorf("bitbucket API returned %s: %s", resp.Status, apiError.Error.Message)
	}

	return fmt.Errorf("bitbucket API returned %s", resp.Status)
}

File: src/scm/bitbucket.ts
import type { SCMClient, PullRequestFile, PRComment } from './types.js';

interface BitbucketConfig {
  workspace: string;
  repoSlug: string;
  token: string;
  baseURL: string;
}

export class BitbucketClient implements SCMClient {
  private config: BitbucketConfig;

  constructor(cfg: BitbucketConfig) {
    this.config = { ...cfg };
    if (!this.config.baseURL) this.config.baseURL = 'https://api.bitbucket.org/2.0';
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = {};
    if (this.config.token) h['Authorization'] = `Bearer ${this.config.token}`;
    return h;
  }

  async pullRequestExists(number: number): Promise<boolean> {
    const url = `${this.config.baseURL}/repositories/${this.config.workspace}/${this.config.repoSlug}/pullrequests/${number}`;
    const resp = await fetch(url, { headers: this.headers() });
    if (resp.status === 404) return false;
    if (!resp.ok) throw new Error(`unexpected status ${resp.status}`);
    return true;
  }

  async getPullRequestDescription(number: number): Promise<{ title: string; body: string }> {
    const url = `${this.config.baseURL}/repositories/${this.config.workspace}/${this.config.repoSlug}/pullrequests/${number}`;
    const resp = await fetch(url, { headers: this.headers() });
    if (!resp.ok) throw new Error(`unexpected status ${resp.status}`);
    const pr = (await resp.json()) as { title: string; summary: { raw: string } };
    return { title: pr.title, body: pr.summary?.raw ?? '' };
  }

  async getPullRequestDiff(number: number): Promise<string> {
    const url = `${this.config.baseURL}/repositories/${this.config.workspace}/${this.config.repoSlug}/pullrequests/${number}/diff`;
    const resp = await fetch(url, { headers: this.headers() });
    if (resp.status === 404) throw new Error(`pull request #${number} not found`);
    if (!resp.ok) throw new Error(`unexpected status ${resp.status}`);
    return resp.text();
  }

  async getPullRequestFiles(number: number): Promise<PullRequestFile[]> {
    const entries = await this.fetchDiffstat(number);
    const fullDiff = await this.getPullRequestDiff(number);
    const patches = splitDiff(fullDiff);

    return entries.map((e) => ({
      filename: e.path,
      status: e.status,
      patch: patches[e.path] ?? '',
      content: '',
    }));
  }

  async getFileContent(path: string, ref: string): Promise<string> {
    if (!ref) ref = 'HEAD';
    const url = `${this.config.baseURL}/repositories/${this.config.workspace}/${this.config.repoSlug}/src/${ref}/${path}`;
    const resp = await fetch(url, { headers: this.headers() });
    if (resp.status === 404) throw new Error(`file "${path}" not found`);
    if (!resp.ok) throw new Error(`unexpected status ${resp.status}`);
    return resp.text();
  }

  async publishComment(number: number, body: string): Promise<void> {
    const url = `${this.config.baseURL}/repositories/${this.config.workspace}/${this.config.repoSlug}/pullrequests/${number}/comments`;
    const resp = await fetch(url, {
      method: 'POST',
      headers: { ...this.headers(), 'Content-Type': 'application/json' },
      body: JSON.stringify({ content: { raw: body } }),
    });
    if (resp.status === 404) throw new Error(`pull request #${number} not found`);
    if (resp.status !== 201) {
      const err = await parseBitbucketError(resp);
      throw new Error(err);
    }
  }

  async getPRComments(number: number): Promise<PRComment[]> {
    const all = await this.fetchAllComments(number);
    return all.filter((c) => !c.path);
  }

  async getPRReviewComments(number: number): Promise<PRComment[]> {
    const all = await this.fetchAllComments(number);
    return all.filter((c) => !!c.path);
  }

  private async fetchAllComments(number: number): Promise<PRComment[]> {
    let url: string | null =
      `${this.config.baseURL}/repositories/${this.config.workspace}/${this.config.repoSlug}/pullrequests/${number}/comments`;
    const allComments: PRComment[] = [];

    while (url) {
      const resp = await fetch(url, { headers: this.headers() });
      if (resp.status === 404) throw new Error(`pull request #${number} not found`);
      if (!resp.ok) throw new Error(`unexpected status ${resp.status}`);

      const page = (await resp.json()) as {
        values: {
          id: number;
          created_on: string;
          content: { raw: string };
          user: { display_name: string };
          inline?: { path: string; to: number };
          parent?: { id: number };
        }[];
        next?: string;
      };

      for (const bc of page.values) {
        allComments.push({
          id: bc.id,
          author: bc.user.display_name,
          body: bc.content.raw,
          createdAt: new Date(bc.created_on),
          path: bc.inline?.path ?? '',
          line: bc.inline?.to ?? 0,
          inReplyTo: bc.parent?.id ?? 0,
        });
      }

      url = page.next ?? null;
    }

    return allComments;
  }

  private async fetchDiffstat(number: number): Promise<{ status: string; path: string }[]> {
    let url: string | null =
      `${this.config.baseURL}/repositories/${this.config.workspace}/${this.config.repoSlug}/pullrequests/${number}/diffstat`;
    const entries: { status: string; path: string }[] = [];

    while (url) {
      const resp = await fetch(url, { headers: this.headers() });
      if (!resp.ok) throw new Error(`unexpected status ${resp.status}`);

      const page = (await resp.json()) as {
        values: { status: string; new: { path: string }; old: { path: string } }[];
        next?: string;
      };

      for (const v of page.values) {
        entries.push({
          status: v.status,
          path: v.new?.path || v.old?.path,
        });
      }

      url = page.next ?? null;
    }

    return entries;
  }
}

function splitDiff(fullDiff: string): Record<string, string> {
  const patches: Record<string, string> = {};
  const chunks = fullDiff.split('diff --git ');

  for (const chunk of chunks.slice(1)) {
    const newlineIdx = chunk.indexOf('\n');
    if (newlineIdx === -1) continue;
    const header = chunk.slice(0, newlineIdx);
    const parts = header.split(/\s+/);
    if (parts.length < 2) continue;
    const filePath = parts[1].replace(/^b\//, '');
    patches[filePath] = 'diff --git ' + chunk;
  }

  return patches;
}

async function parseBitbucketError(resp: Response): Promise<string> {
  try {
    const body = (await resp.json()) as { error?: { message?: string } };
    if (body.error?.message) {
      return `bitbucket API returned ${resp.status} ${resp.statusText}: ${body.error.message}`;
    }
  } catch {
    // ignore parse errors
  }
  return `bitbucket API returned ${resp.status} ${resp.statusText}`;
}
